pkg/server/plugin/upstreamauthority/certmanager/certmanager_test.go

```diff
@@ -182,15 +181,10 @@ func Test_MintX509CA(t *testing.T) {
 			require.NoError(t, err)
 
 			go func() {
-				var cr *cmapi.CertificateRequest
-				for cr == nil {
-					time.Sleep(time.Nanosecond)
-					crList := &cmapi.CertificateRequestList{}
-					require.NoError(t, cmclient.List(context.TODO(), crList))
-					if len(crList.Items) > 0 {
-						cr = &crList.Items[0]
-					}
-				}
+				<-crCreated
+				crList := &cmapi.CertificateRequestList{}
+				require.NoError(t, cmclient.List(context.TODO(), crList))
```

Comment:
These `require`'s should not happen in this goroutine. Perhaps, with the hook, this goroutine is no longer needed? The main goroutine can just wait for crCreated and do the assertions directly?

Reply:
I believe this second goroutine is still required because the `certmanager` implementation of `MintX509CA` polls the cert-manager API until it has the `Status.Certificate` field populated. Based on my understanding of the [API docs](https://cert-manager.io/docs/reference/api-docs/#cert-manager.io/v1.CertificateRequestStatus), cert-manager should populate that field asynchronously after the certificate request has been submitted.

When the plugin is using the fake K8s client, the tests have to populate the `Status.Certificate` field. Since `MintX509CA` is a server-side streaming API, the main goroutine of the test will be blocked on the `r.Recv()` until the call to `test.updateCR(t, cr)` takes place, which has to happen after the client listen on the stream since `MintX509CA` is what creates the certificate request in the K8s API server.

For that reason, I think this update task has to take place in a separate goroutine based on the current test design.

Reply:
Is there a specific concern you have with assertions in spawned goroutines in tests?

Reply:
`require`'s don't work as expected on spawned goroutines and will cause a panic. Further, `assert'`s also require careful diligence since if they assert after the test has finished, it will also cause a panic.

We either need to 1) convert these require's into assert's, or 2) do the list operation in the goroutine but send the results down a channel that the main goroutine can pick up and assert.

Reply:
Got it, that makes sense. Converting these to asserts seems pretty straightforward.